rfcs/proposed/task_group_dynamic_dependencies/README.md

```diff
@@ -0,0 +1,378 @@
+
+        for (tbb::task_handle& p : users::find_predecessors(work_id)) {
+            new_task.add_predecessor(p);
+        }
+
+        tg.run(new_task);
+    }
+
+While the graph, as shown below, is simple, the completion status of the predecessors
+is unknown. Therefore, for ease of use, `task_handle` should be usable as a dependency
+regardless of state of the task it represents. Any predecessor that is already
+completed when it is added as a predecessor will not delay the start of the dependent
+task.
+
+<img src="unknown_states.png" width=400>
+
+#### Recursive Decomposition
+
+This example is a version of merge-sort (with many of the details left out). 
+Assume an initial task executes the function shown below as its body, and the
+function implements that task, and also serves as the body for the recursively
+decomposed pieces. The range of the sequence is defined by `b` (beginning) and
+end `e` (end). Most of the implementation details of merge sort are abstracted
+into the following helper functions: `users::do_serial_sort`, `users::create_left_range`,
+`users::create_right_range`, and `users::do_merge`.
+
+    template<typename T>
+    void merge_sort(tbb::task_group& tg, T b, T e) {
+        if (users::range_is_too_small(b, e)) {
+            // base-case when range is small
+            users::do_serial_sort(b, e);
+        } else {
+            // calculate left and right ranges
+            T lb, le, rb, re;
+            users::create_left_range(lb, le, b, e);
+            users::create_right_range(rb, re, b, e);
+
+            // create the three tasks
+            tbb::task_handle sortleft =
+                tg.defer([lb, le, &tg] {
+                    merge_sort(tg, lb, le);
+                });
+            tbb::task_handle sortright =
+                tg.defer([rb, re, &tg] {
+                    merge_sort(tg, rb, re);
+                });
+            tbb::task_handle merge =
+            tg.defer([rb, re, &tg] {
+                    users::do_merge(tg, lb, le, rb, re);
+                });
+
+            // add predecessors for new merge task
+            merge.add_predecessor(sortleft);
+            merge.add_predecessor(sortright);
+
+            // insert new subgraph between currently executing
+            // task and its successors
+            tbb::task_group::transfer_successors_to(merge);
+
+            tg.run(sortleft);
+            tg.run(sortright);
+            tg.run(merge);
+        }
+    }
+
+This task tree matches the one shown earlier for merge-sort. 
+
+## Open Questions in Design
+
+Some open questions that remain:
+
+- Are the suggested APIs sufficient?
+- Should we add a function to adds more than one predecessor as single call, such as `add_predecessors`?
+- Should we add functions that merge creation and definition of predecessor tasks, such as
+`template <typename Func> add_predecessor(Func&& f);`.
+- Are there additional use cases that should be considered that we missed in our analysis?
```

Comment:
I can suggest a few more potentially interesting cases:
- a wavefront, where tasks have multiple predecessors and multiple successors;
- a two-stage parallel scan (reduce-then-scan), with left-to-right propagation of the accumulated prefix sum;
- non-trivial divide-and-conquer patterns, such as Matteo Frigo's algorithm for N-Body problem (see e.g. https://dspace.mit.edu/bitstream/handle/1721.1/122680/6-172-fall-2010/contents/projects/MIT6_172F10_proj4_1.pdf)

These seem to be interesting examples of "dynamic task graphs that are not trees", as stated in the introduction. I do not suggest that we must support these patterns, but it is interesting to see if we could, and what would be needed for that.

Reply:
Yes, more example would be good. In particular, we need to see if the very limited ways to add and modified predecessors is sufficient. And how lifetimes of task handles will be managed.